x/supplier/keeper/msg_server_stake_supplier.go

```diff
@@ -3,15 +3,95 @@ package keeper
 import (
 	"context"
 
+	sdkerrors "cosmossdk.io/errors"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sharedtypes "pocket/x/shared/types"
 	"pocket/x/supplier/types"
 )
 
-func (k msgServer) StakeSupplier(goCtx context.Context, msg *types.MsgStakeSupplier) (*types.MsgStakeSupplierResponse, error) {
+func (k msgServer) StakeSupplier(
+	goCtx context.Context,
+	msg *types.MsgStakeSupplier,
+) (*types.MsgStakeSupplierResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// TODO: Handling the message
-	_ = ctx
+	logger := k.Logger(ctx).With("method", "StakeSupplier")
+	logger.Info("About to stake supplier with msg: %v", msg)
+
+	// Check if the supplier already exists or not
+	var err error
+	var coinsToDelegate sdk.Coin
+	supplier, isSupplierFound := k.GetSupplier(ctx, msg.Address)
+	if !isSupplierFound {
+		logger.Info("Supplier not found. Creating new supplier for address %s", msg.Address)
+		if err = k.createSupplier(ctx, &supplier, msg); err != nil {
+			return nil, err
+		}
+		coinsToDelegate = *msg.Stake
+	} else {
+		logger.Info("Supplier found. Updating supplier for address %s", msg.Address)
+		currSupplierStake := *supplier.Stake
+		if err = k.updateSupplier(ctx, &supplier, msg); err != nil {
+			return nil, err
+		}
+		coinsToDelegate = (*msg.Stake).Sub(currSupplierStake)
+	}
+
+	// Retrieve the address of the supplier
+	supplierAddress, err := sdk.AccAddressFromBech32(msg.Address)
+	if err != nil {
+		logger.Error("could not parse address %s", msg.Address)
+		return nil, err
+	}
+
+	// Send the coins from the supplier to the staked supplier pool
+	err = k.bankKeeper.DelegateCoinsFromAccountToModule(ctx, supplierAddress, types.ModuleName, []sdk.Coin{coinsToDelegate})
+	if err != nil {
+		logger.Error("could not send %v coins from %s to %s module account due to %v", coinsToDelegate, supplierAddress, types.ModuleName, err)
+		return nil, err
+	}
+
+	// Update the Supplier in the store
+	k.SetSupplier(ctx, supplier)
+	logger.Info("Successfully updated supplier stake for supplier: %+v", supplier)
 
 	return &types.MsgStakeSupplierResponse{}, nil
 }
+
+func (k msgServer) createSupplier(
```

Comment:
NIT: Here and in the others I think its strange passing in the value and overwriting it. As we are assigning anyway I think instead of returning error (doesn't ever happen) we should return the `Supplier` and overwrite the variable we pass in as an arg. See [1]

[1] https://github.com/pokt-network/poktroll/blob/f520215b3c08b800ebcac7e1f6375ee312be6f20/x/gateway/keeper/msg_server_stake_gateway.go#L27